src/mongocxx/test/gridfs/bucket.cpp

```diff
@@ -1059,4 +1056,57 @@ TEST_CASE("gridfs download large file", "[gridfs::bucket]") {
 
     REQUIRE(total_bytes_read == length);
 }
+
+TEST_CASE("gridfs automatic index creation", "[gridfs::uploader] [gridfs::downloader]") {
+    using bsoncxx::builder::basic::document;
+    using bsoncxx::builder::basic::sub_array;
+
+    instance::current();
+    client client{uri{}};
+    database db = client["gridfs_index_creation"];
+    auto bucket = db.gridfs_bucket();
+
+    const size_t file_size = 100;
+    std::array<std::uint8_t, file_size> to_write, to_read;
+    std::iota(begin(to_write), end(to_write), 0);
+
+    auto uploader = bucket.open_upload_stream("test_file");
+
+    uploader.write(to_write.data(), file_size);
+    auto result = uploader.close();
+
+    auto downloader = bucket.open_download_stream(result.id());
+    auto bytes_read = downloader.read(to_read.data(), file_size);
+    REQUIRE(bytes_read == file_size);
+
+    SECTION("default indexes created") {
+        auto bucket_collection_name = GENERATE(values({"fs.chunks", "fs.files"}));
+        auto indexes = db[bucket_collection_name].list_indexes();
+        REQUIRE(std::distance(indexes.begin(), indexes.end()) == 2);
+    }
+
+    SECTION("fs.chunks 'files_id' and 'n' compares by numeric value") {
```

Comment:
Per discussion, instead of attempting to create conflicting indexes after the gridfs bucket, create the indexes first (with varying the value between `1` and `1.0`, similar to the libmongoc test). Then assert that GridFS did not create additional indexes. 

We want to test the index comparison logic in GridFS. Creating the indexes first, then checking that GridFS did not create additional indexes, verifies that GridFS correctly detected that the indexes exist.

Note, looking at `create_indexes_if_nonexistent`, IIUC GridFS isn't checking for indexes at all (just that the collections are empty). So I expect tests would fail, and that `create_indexes_if_nonexistent` will need to be updated.

Reply:
Doing it this way worked!! Now I create the index for `chunks` and `files` manually before uploading and downloading arbitrary data and ensure that no more indexes were created. I also updated the `TEST_CASE` and `SECTION` description to explicitly state that this is being tested. Also, TIL, that this code works...
```
TEST_CASE (...) {
    int i = 0;
    REQUIRE(i==0);
    SECTION(...) {
        i++;
        REQUIRE(i==1);
    }
    SECTION(...) {
        i++;
        REQUIRE(i==1);
    }
    REQUIRE(i==1);
}
```
This is what I modeled my code after.

Reply:
Nice!

Re: my note about `create_indexes_if_nonexistent`, it really does not check whether indexes exist before attempting to create them. But I think that's ok, because the server will not return an error when attempting to create the same index twice.

Here's what happens creating the same index twice:
```
db> db.runCommand({"createIndexes": "coll", indexes: [{"name": "test", key: {test: 1}}]})
db> db.runCommand({"createIndexes": "coll", indexes: [{"name": "test", key: {test: 1}}]})
{
	"numIndexesBefore" : 4,
	"numIndexesAfter" : 4,
	"note" : "all indexes already exist",
	"ok" : 1,
	"$clusterTime" : {
		"clusterTime" : Timestamp(1594661966, 1),
		"signature" : {
			"hash" : BinData(0,"AAAAAAAAAAAAAAAAAAAAAAAAAAA="),
			"keyId" : NumberLong(0)
		}
	},
	"operationTime" : Timestamp(1594661966, 1)
}
```

The second reply is not an error ("ok": 1). That seems ok to me. The GridFS spec does not say how to check for indexes, pymongo seems to rely on the same behavior, and even if we did check for indexes explicitly it'd be the same number of commands.